### DISCUSSION TITLE
Not Working  beforeCreate()

### DISCUSSION THREAD

**Discussion by paingchan** [filamentphp/filament#12222] [Help]
### Package

Panel builder

### Package Version

v3.0.0

### How can we help you?

`    protected function beforeCreate(): void
{
    Log::info('beforeCreate method called');
    $formData = $this->getFormData(); // Get the form data
    $orderItems = $formData['order_items']; // Get the order items from the form

    foreach ($orderItems as $orderItem) {
        $inventoryId = $orderItem['inventory_id'];
        $quantity = (int)$orderItem['quantity'];
        $inventory = Inventory::find($inventoryId);

        if ($inventory && $quantity <= $inventory->quantity) {
            $inventory->quantity -= $quantity;
            $inventory->save();
            Notification::make()
            ->success()
            ->title('Quantity Update on Inventory')
            ->body('The quantity update to......')
            ->send(); 
        } 
    }
}`

`             TableRepeater::make('Order_items')
                        ->relationship(name :'order_items')
                        ->reorderableWithDragAndDrop(false)
                        ->addable(false)
                        ->columnSpan(10)
                        ->required()
                        ->headers([
                            Header::make('Barcode')->width('150px'),
                            Header::make('Warehouse')->width('150px'),
                            Header::make('name')->width('150px'),
                            Header::make('qty')->width('150px'),
                            Header::make('price')->width('150px'),
                            Header::make('Total')->width('150px'),
                        ])
                        ->schema([
TextInput::make('inventory_id')
                                ->readOnly()
                                //->hidden()
                                ->label('Product')
                                ->disabledOn('edit')
                                //->live()
                                ->required(), 
TextInput::make('quantity')
                                ->live()
                                ->required()
                                ->integer()
                                ->label('Qty')
                               ->afterStateUpdated(function (callable $set, Get $get, ?string $state) {
                                    $inventoryId = $get('inventory_id');
                                    $quantity = (int)$get('quantity');
                                    $inventory = Inventory::find($inventoryId);

                                    if ($inventory && $quantity > $inventory->quantity) {
                                        Notification::make()
                                            ->danger()
                                            ->title('Quantity Exceeded')
                                            ->body('The quantity cannot exceed the available stock')
                                            ->send();
                                            
                                        $set('quantity', $inventory->quantity);
                                    }
                                }),`

**Comment by micraux** [comment]
Hello,

Same problem for me here, in a classic "_order / order_lines_" create situation.

I need to calculate the _invoice_number_ before saving the order in the database.

So I just tried this in my create page:
```php
// CreateOrder.php
protected function beforeCreate(): void
{
    $this->data['invoice_number'] = 999;
}
```

This doesn't work!

If I `dd($this->data)`, I can see the _invoice_number_ field populated with the _999_ value, but the column is empty on the database.

Am I missing something?

**Comment by micraux** [comment]
Hey @paingchan, I came to a solution, using `handleRecordCreation()` instead of `beforeCreate()`.

First, you have to add `->dehydrated(true)` in your Repeater ([see comment here](https://github.com/filamentphp/filament/discussions/8815#discussioncomment-7165972)) but be careful: put it directly under `->relationship(name :'order_items')` because I tried putting it lower and it wasn't working properly.

Then you will be able to access the repeater data in `handleRecordCreation()`

**Comment by Lisfe00** [comment]
You can use mutateFormDataBeforeSave(array $data): array in Edit page 
and use mutateFormDataBeforeCreate(array $data): array in Create page
